### PR TITLE
Try applying national prefixes when parsing phone numbers.

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### CODE REVIEW

File: src/Contacts/PhoneNumber.m [hunk 1]
```diff
@@ -34,8 +46,8 @@ + (PhoneNumber *)phoneNumberFromText:(NSString *)text andRegion:(NSString *)regi
     }
 
     PhoneNumber *phoneNumber = [PhoneNumber new];
```
Comment (michaelkirk) [inline src/Contacts/PhoneNumber.m]: take it or leave it: 

But as long as we're touching this, how about `return [[PhoneNumber alloc] initWithNBPhoneNumber:e164:]` and making the properties readonly.

File: src/Contacts/PhoneNumber.m [hunk 2]
```diff
@@ -119,9 +131,91 @@ + (PhoneNumber *)tryParsePhoneNumberFromUserSpecifiedText:(NSString *)text {
...
++ (NSString *)nationalPrefixTransformRuleForClientPhoneNumber:(NSString *)clientPhoneNumber
+{
+    if (clientPhoneNumber.length < 1) {
+        return nil;
+    }
+    static NSString *result = nil;
+    static NSString *cachedClientPhoneNumber = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        NSNumber *localCallingCode = [[PhoneNumber phoneNumberFromE164:clientPhoneNumber] getCountryCode];
```
Comment (michaelkirk) [inline src/Contacts/PhoneNumber.m]: It took me a minute to realize what was going on here.

Can you add a comment above the assert to the affect of

```
// clientPhoneNumber is the local users phone number and should never change
```

File: src/Contacts/PhoneNumber.m [hunk 3]
```diff
@@ -211,30 +305,31 @@ + (PhoneNumber *)tryParsePhoneNumberFromE164:(NSString *)text {
...
     NBPhoneNumberUtil *phoneUtil = [PhoneNumberUtil sharedUtil].nbPhoneNumberUtil;
 
     NSError *formatError = nil;
-    NSString *pretty = [phoneUtil format:phoneNumber numberFormat:NBEPhoneNumberFormatINTERNATIONAL error:&formatError];
+    NSString *pretty =
+        [phoneUtil format:self.phoneNumber numberFormat:NBEPhoneNumberFormatINTERNATIONAL error:&formatError];
 
     if (formatError != nil)
-        return e164;
+        return self.e164;
```
Comment (michaelkirk) [inline src/Contacts/PhoneNumber.m]: Thanks for the cleanup!

As long as you're touching this (or clang format is...) can you add the braces?

File: src/Contacts/PhoneNumber.m [hunk 4]
```diff
@@ -119,9 +131,91 @@ + (PhoneNumber *)tryParsePhoneNumberFromUserSpecifiedText:(NSString *)text {
...
+        if (![normalizedText containsString:@"$"]) {
+            [result addObjectsFromArray:[self tryParsePhoneNumbersFromsNormalizedText:normalizedText
+                                                                    clientPhoneNumber:clientPhoneNumber]];
+        }
+    }
+
+    return [result copy];
+}
+
++ (NSArray<PhoneNumber *> *)tryParsePhoneNumbersFromsNormalizedText:(NSString *)text
```
Comment (michaelkirk) [inline src/Contacts/PhoneNumber.m]: typo in method name?

> tryParsePhoneNumbersFrom**s**NormalizedText

**Review by michaelkirk — APPROVED** [review]
LGTM

This code is a little gnarly, but that's because phone number parsing is gnarly.

If history tells us anything, it's that there will be more issues with parsing phone numbers in the future.

It seems like some unit tests would be straight forward, and also potentially useful for catching regressions as we continue to hammer this code back and forth.

File: src/Contacts/PhoneNumber.m (hunk 1 above)
Comment (charlesmchen) [reply src/Contacts/PhoneNumber.m]: Done.

File: src/Contacts/PhoneNumber.m (hunk 2 above)
Comment (charlesmchen) [reply src/Contacts/PhoneNumber.m]: Done.

File: src/Contacts/PhoneNumber.m (hunk 4 above)
Comment (charlesmchen) [reply src/Contacts/PhoneNumber.m]: Done.

File: src/Contacts/PhoneNumber.m (hunk 3 above)
Comment (charlesmchen) [reply src/Contacts/PhoneNumber.m]: Done.

**Comment by charlesmchen** [comment]
> It seems like some unit tests would be straight forward, and also potentially useful for catching regressions as we continue to hammer this code back and forth.

I'd love to do this... after we fix the tests?  

* https://github.com/WhisperSystems/SignalServiceKit/pull/284
* https://github.com/WhisperSystems/SignalServiceKit/pull/244

**Comment by michaelkirk** [comment]
>> It seems like some unit tests would be straight forward, and also potentially useful for catching >>regressions as we continue to hammer this code back and forth.
>
> I'd love to do this... after we fix the tests?
>
> #284
> #244

FYI - tests are now fixed.

File: src/Contacts/PhoneNumber.m [hunk 5]
```diff
@@ -32,10 +56,7 @@ + (PhoneNumber *)phoneNumberFromText:(NSString *)text andRegion:(NSString *)regi
         return nil;
     }
 
-    PhoneNumber *phoneNumber = [PhoneNumber new];
-    phoneNumber->phoneNumber = number;
-    phoneNumber->e164        = e164;
-    return phoneNumber;
+    return [[PhoneNumber alloc] initWithPhoneNumber:number e164:e164];
```
Comment (michaelkirk) [inline src/Contacts/PhoneNumber.m:59]: 👏 👏 👏

File: tests/Contacts/PhoneNumberTest.m [hunk 6]
```diff
@@ -5,10 +5,29 @@
...
 
+@interface NSArray (PhoneNumberTest)
+
+@end
+
+#pragma mark -
+
+@implementation NSArray (PhoneNumberTest)
+
+- (id)test_firstObject
```
Comment (michaelkirk) [inline tests/Contacts/PhoneNumberTest.m:16]: Am I missing something or is this exactly the same as the built in `[myArray firstObject]`

File: tests/Contacts/PhoneNumberTest.m [hunk 7]
```diff
@@ -59,4 +78,72 @@ - (void)testTryParsePhoneNumberFromUserSpecifiedTextRemovesAnyFormatting {
...
+        [result addObject:phoneNumber.toE164];
+    }
+    return result;
+}
+
+- (void)testTryParsePhoneNumbersFromsUserSpecifiedText_SimpleUSA
+{
+    NSArray<NSString *> *parsed =
+        [self unpackTryParsePhoneNumbersFromsUserSpecifiedText:@"323 555 1234" clientPhoneNumber:@"+13213214321"];
+    XCTAssertLessThanOrEqual(1, parsed.count);
```
Comment (michaelkirk) [inline tests/Contacts/PhoneNumberTest.m:97]: I find some of the XCT macros to be a bit hard to read - the compare functions among them. I guess I'm just used to using comparators in infix notation.

So, I think this is equivalent to `XCTAssert(1 <= parsed.count)`
But I think it'd be a little more idiomaticly expressed in reverse:  `XCTAssert(parsed.count >= 1)`
Which I guess would be: `XCTAssertGreaterThanOrEqual(parsed.count, 1)`
But honestly I prefer the: `XCTAssert(parsed.count >= 1)`

I know it's all the same, so feel free to ignore, but I had to do a triple take when reading this.

File: tests/Contacts/PhoneNumberTest.m [hunk 8]
```diff
@@ -59,4 +78,72 @@ - (void)testTryParsePhoneNumberFromUserSpecifiedTextRemovesAnyFormatting {
...
+
+    parsed = [self unpackTryParsePhoneNumbersFromsUserSpecifiedText:@"8341639157" clientPhoneNumber:@"+528341639144"];
+    XCTAssertLessThanOrEqual(1, parsed.count);
+    XCTAssertTrue([parsed containsObject:@"+528341639157"]);
+
+    parsed = [self unpackTryParsePhoneNumbersFromsUserSpecifiedText:@"341639157" clientPhoneNumber:@"+528341639144"];
+    XCTAssertLessThanOrEqual(1, parsed.count);
+    // The parsing logic should try adding Mexico's national prefix for cell numbers "1"
+    // after the country code.
+    XCTAssertTrue([parsed containsObject:@"+521341639157"]);
```
Comment (michaelkirk) [inline tests/Contacts/PhoneNumberTest.m:138]: should we also:   
```
XCTAssertTrue([parsed containsObject:@"+52341639157"]);
```

(without the inserted `1`)

**Review by michaelkirk — APPROVED** [review]
LGTM 

My only concern would be if this substantially increases contact parsing time.

**Comment by charlesmchen** [comment]
> My only concern would be if this substantially increases contact parsing time.

It should only affect Argentina and Mexico and in those cases it should worst-case double contact parsing time.

File: tests/Contacts/PhoneNumberTest.m (hunk 6 above)
Comment (charlesmchen) [reply tests/Contacts/PhoneNumberTest.m:16]: That's so weird. I thought they only had `lastObject`. Nice.

File: tests/Contacts/PhoneNumberTest.m (hunk 7 above)
Comment (charlesmchen) [reply tests/Contacts/PhoneNumberTest.m:97]: Done.

File: tests/Contacts/PhoneNumberTest.m (hunk 8 above)
Comment (charlesmchen) [reply tests/Contacts/PhoneNumberTest.m:138]: Done.

**Comment by charlesmchen** [comment]
Migrated to https://github.com/WhisperSystems/Signal-iOS/pull/2324

**Pull request closed without merging**